iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/functional_interface.inl

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_CXX_FUNCTIONAL_INTERFACE_INL
+#define IOX_HOOFS_CXX_FUNCTIONAL_INTERFACE_INL
+
+namespace iox
+{
+namespace cxx
+{
+namespace internal
+{
+template <typename T>
+inline void Expect<T>::expect(const char* const msg) const noexcept
+{
+    if (!(*static_cast<const T*>(this)))
+    {
+        std::cout << msg << std::endl;
+        Ensures(false);
+    }
+}
+
+template <typename T, typename ValueType>
+inline ValueType& ExpectWithValue<T, ValueType>::expect(const char* const msg) & noexcept
+{
+    T* upcastedThis = static_cast<T*>(this);
+
+    if (!(*upcastedThis))
+    {
+        std::cout << msg << std::endl;
+        Ensures(false);
+    }
+
+    return upcastedThis->value();
+}
+
+template <typename T, typename ValueType>
+inline const ValueType& ExpectWithValue<T, ValueType>::expect(const char* const msg) const& noexcept
+{
+    return const_cast<const ValueType&>(const_cast<ExpectWithValue<T, ValueType>*>(this)->expect(msg));
+}
+
+template <typename T, typename ValueType>
+inline ValueType&& ExpectWithValue<T, ValueType>::expect(const char* const msg) && noexcept
+{
+    return std::move(this->expect(msg));
+}
+
+template <typename T, typename ValueType>
+inline const ValueType&& ExpectWithValue<T, ValueType>::expect(const char* const msg) const&& noexcept
+{
+    return const_cast<const ValueType&&>(std::move(const_cast<ExpectWithValue<T, ValueType>*>(this)->expect(msg)));
+}
+
+template <typename T, typename ValueType>
+inline ValueType ValueOr<T, ValueType>::value_or(const ValueType& value) const noexcept
```

Comment:
No need to change, but I think readability would improve when you rename `value` to `alternative` like you did in the header file.